client/src/GameState.ts

```diff
@@ -174,43 +174,38 @@ export class GameState {
 
     public fromBoardState(boardState: BoardState): Array<Monomino> {
         const needRedraw = [];
-        if (this.playerId && this.playerId > 0) {
-            // rotate remote player0 board to local board
-            const ccRotations = 4 - ((this.playerId || 0) % 4);
-            const localBoard = boardState.map((row) => row.slice());
-            for (let row = 0; row < BOARD_SIZE; row++) {
-                for (let col = 0; col < BOARD_SIZE; col++) {
-                    const [newRow, newCol] = Tetromino.rotateCoords(
-                        [row, col],
-                        BOARD_SIZE,
-                        ccRotations
-                    );
-                    localBoard[newRow][newCol] = boardState[row][col];
-                    const monominoState = boardState[row][col];
+        const ccRotations = 4 - ((this.playerId || 0) % 4);
+
+        for (let row = 0; row < BOARD_SIZE; row++) {
+            for (let col = 0; col < BOARD_SIZE; col++) {
+                const [newRow, newCol] = Tetromino.rotateCoords(
+                    [row, col],
+                    BOARD_SIZE,
+                    ccRotations
+                );
+                // the future state of the monomino at [row, col]
+                let monominoState = boardState[row][col];
+                if (this.playerId && this.playerId > 0) {
+                    // rotate from player0 view
+                    monominoState = boardState[row][col];
```

Comment:
Double set of `monominoState = boardState[row][col];` ?

Reply:
Also, I don't think you even need the playerId check before updating the monomino's position with the new coordinates, you just may set it as the same thing on occasion. You could just include that update in the else section below before calling `updateFromMonominoState`

Reply:
Thanks, I had brain outage on this big for loop :)